### ISSUE TITLE
About Training Helper Problem

### ISSUE THREAD

**Issue by azraelkuan** [Rayhane-mamah/Tacotron-2#48]
i have rewrite the `feeder.py` to the lastest `tf.dataset`, but i find a problem in your code:
https://github.com/Rayhane-mamah/Tacotron-2/blob/d375c67fbfda2989294ecb4ec9892762055d8277/tacotron/models/helpers.py#L129-L132

the `next_inputs` in the `helper.py` depends on the` tfr`, but if time is the final step, as i konw, `len(stop_tokens) = len(targets) + 1`, so if the ratio < `tf.random_uniform`, the target will not have the indice `len(stop_tokens)`, it will cause a index error.
some solutions like this:
https://github.com/tensorflow/tensorflow/blob/cfd0ea3bfb85d92cdb32760ea024f1e38618d717/tensorflow/contrib/seq2seq/python/ops/helper.py#L248-L250
this will return a zero tensor if finish is true.
thanks, may be it is a bug in my code.

**Comment by Rayhane-mamah** [comment]
Hello @azraelkuan, thanks for reaching out!

Oh yeah I forgot to delete this part as it is pretty much useless now:
https://github.com/Rayhane-mamah/Tacotron-2/blob/d375c67fbfda2989294ecb4ec9892762055d8277/tacotron/models/helpers.py#L121-L123

It is what's mainly causing your error because model isn't seeing the stop token in all batch samples despite the targets being finished, since it is useless I will delete it for future commit and always use this instead:

https://github.com/Rayhane-mamah/Tacotron-2/blob/d375c67fbfda2989294ecb4ec9892762055d8277/tacotron/models/helpers.py#L124-L126

However, this is only crashing for your because you are setting `len(stop_tokens) = len(targets) + 1 `which is kind of wrong.. In my project `len(stop_token) = len(targets)`, and token_targets are all zeros but the last one (when no extra padding is added). i.e: even without paddings, the token of last time step` must be 1`.

I just want to bring your attention to this detail in case you missed it while changing the `feeder.py`:

https://github.com/Rayhane-mamah/Tacotron-2/blob/d375c67fbfda2989294ecb4ec9892762055d8277/tacotron/feeder.py#L191

When first creating the token targets, I intentionally create vectors of length `len(targets) - 1` so that padding automatically adds a 1 for length matching, and the rest of 1s for padding.

That explains why I add 1 to the length of `stop_tokens:`

https://github.com/Rayhane-mamah/Tacotron-2/blob/d375c67fbfda2989294ecb4ec9892762055d8277/tacotron/feeder.py#L216

Once you get your project up and running, could you report if `tf.dataset` is faster than our approach for feeding data to the model? (most importantly to the GPU) I am suspecting that our feeder pipeline is causing a slowdown that makes the GPU-Util drop between training steps. Thank a lot!

I hope this was helpful, if the problem persists, feel free to share more about how your feeder works, and we'll try figuring things out.

**Comment by azraelkuan** [comment]
thanks! i will test the dataset.

**Comment by azraelkuan** [comment]
another question, how many seconds per step in your feeder?

**Comment by Rayhane-mamah** [comment]
Usually, a training step takes about 1.7 seconds when using r=1 and batch_size=32.

The feeder generates 32 batches of 32 samples in 1.5 seconds, but the problem isn't during generation, it's more likely during feeding those batches to the model.

As a note, the GPU-Util between steps might be due to something else, but my first guess if the data feedings pipeline, are you experiencing such slowdown with tf.dataset too? 

Thanks for the feedback

**Comment by a3626a** [comment]
I have also rewritten feeder using tf.Dataset. I found that :
1) gpu utilization
  I rewrote feeder to increase gpu utilization. However, there are no dramatic changes. After profiling the model using tf.Profiler, I concluded that low gpu utilization is due to intrinsic sequential property of the model. gpu utilization gets lower with smaller reduction factor.

2) batching
  I optimized batching algorithm using `tf.contrib.data.bucket_by_sequence_length`. In this repository, it is done by sorting batched group of samples by length of spectrograms. **Be careful! `bucket_by_sequence_length` slows down getting alignment**. I think it is caused by similar reason as masking padding of encoder, decoder slowing down getting alignment.
